### ISSUE TITLE
Shiny Server stops loading for LDAP users with bigger X-Groups 

### ISSUE THREAD

**Issue by erdalkucukk** [rstudio/shiny-server#327]
Hi, 

We have a shiny-server pro which works behind nginx. Authentication is done via LDAP and nginx forwards the x-users and x-groups information to shiny, which checks the authentication.

Until recently, this setup worked fine but lately we changed the number of groups a user is member of  and the server started to creash for users that has many group memberships. 

Initially nginx created the issue with login and as a solution we added a proxy buffering which fixed that issue. 

nginx.conf

```
....
  map $http_upgrade $connection_upgrade {
      default upgrade;
      ''      close;
    }

server {
    listen       80;

....
      proxy_buffering on;
      proxy_buffer_size 512k;
      proxy_buffers 4 512k;
      proxy_busy_buffers_size 512k;

    location / {
      auth_request /_auth;

      # if successful, get x-user and x-groups
      **auth_request_set $user $upstream_http_x_user;
      auth_request_set $groups $upstream_http_x_groups;**

      # pass user and groups to shiny server
      proxy_set_header X-USER $user;
      proxy_set_header X-GROUPS $groups;

      # proxy to shiny server
      proxy_pass http://shiny;

      proxy_http_version 1.1;
      proxy_set_header Upgrade $http_upgrade;
      proxy_set_header Connection "upgrade";
      proxy_set_header X-Real-IP $remote_addr;
      proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
      proxy_set_header Host $http_host;
      proxy_set_header X-NginX-Proxy true;
      proxy_redirect off;
    }

    location /admin {
      auth_request /_auth;
      **auth_request_set $user $upstream_http_x_user;**

      rewrite ^/admin/(.*)$ /$1 break;

      proxy_set_header X-USER $user;
      proxy_set_header X-GROUPS $user;
      proxy_pass http://shiny-admin;
      proxy_http_version 1.1;
      proxy_set_header Upgrade $http_upgrade;
      proxy_set_header Connection "upgrade";
      proxy_set_header X-Real-IP $remote_addr;
      proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
      proxy_set_header Host $http_host;
      proxy_set_header X-NginX-Proxy true;
      proxy_redirect off;
    }
```
I made the test for two users, one has access to 5 groups and no admin rights, and second user with membership to near 50 groups and have admin rights. When I log in with these users test user with less membership works fine. But with the user that has big upstream x-groups info after the log in application breaks and execution is halted message is printed in application logs, with a gray shiny screen hang in the initalization. However, the same users is able to see the shiny server pro admin page, which doesn't check the x-group value, but only the username in the request. 

shiny.conf
```
run_as shiny;

auth_proxy X-USER X-GROUPS;

disable_protocols websocket xdr-streaming xhr-streaming iframe-eventsource iframe-htmlfile;

log_as_user true;

access_log /var/log/shiny-server/access.log combined;

preserve_logs true;

app_init_timeout 300;

server {
  listen 3838 127.0.0.1;
```

shiny-server.log

```
[2017-11-20 14:09:09.269] [DEBUG] shiny-server - Creating a new robust connection: yQG3pBzN0AGMe7cLkR
[2017-11-20 14:09:09.270] [TRACE] shiny-server - Accepted SockJS connection for /tv2/__sockjs__/n=yQG3pBzN0AGMe7cLkR/t=65514ad311b02674d2f32fd8a3a668dc/w=e842c95deca7b81393a2df28e472b336b95882c40bde3e6e/s=0/700/1km_pnh0/xhr
**[2017-11-20 14:09:09.351] [TRACE] shiny-server - Attempted to auth a web socket. Closing.**
**[2017-11-20 14:09:14.483] [TRACE] shiny-server - Timeout expired. Killing process.**

```
Above is the error code we receive for a user  that gets the error. Another user with less membership works and don't print the last two lines. And earlier ldap configuration, the same user was working fine. 

Is it because shiny can't handle big chunks of ldap groups or the nginx proxy passes wrong information for the users such that? How can I fix the issue and where should I look more rather than proxy buffer sizes or http websocket settings?

I tried to add logging information that prints the groups shiny receives in the application but that only works for the succcessful one, not for the one that crashes the app.

I couldn't manage to listen the communication between nginx and shiny-server if the proxy sends correct infromation.

**Comment by erdalkucukk** [comment]
Additionally when I listen the http port with tcpdump and check with wireshark, I see the both successful and not successful ones access to shiny, gets redirected to the page and after the redirection they both send cookies and session information. During this stage, the successful one sends a cookie with truncated session information, and while other one seems like not sending it.

It seems shiny sends a reassembled TCP package since the size is too big, and it breaks the cookie information, thus doesn't include the session info. It is my guess. 

Unsuccessful setup with # of X-groups as 72
![image](https://user-images.githubusercontent.com/33837471/33542436-587fa29a-d8d3-11e7-9fbb-9dc0783ce069.png)
![image](https://user-images.githubusercontent.com/33837471/33542467-75134470-d8d3-11e7-9ed5-6c0454baa805.png)
![image](https://user-images.githubusercontent.com/33837471/33542544-c61e8870-d8d3-11e7-8a1d-c4791ec3ecfa.png)
And the next message
![image](https://user-images.githubusercontent.com/33837471/33542513-a2fd25d6-d8d3-11e7-8fb6-53586ea7f7fd.png)

Successful setup with  # of X-groups as 11
![image](https://user-images.githubusercontent.com/33837471/33542586-f67372ec-d8d3-11e7-830b-fa9b7ea3f933.png)
![image](https://user-images.githubusercontent.com/33837471/33542566-dcb24400-d8d3-11e7-8568-d6b3034c2818.png)
![image](https://user-images.githubusercontent.com/33837471/33542603-0e2898f4-d8d4-11e7-9d7c-1b0ce7dc807e.png)

Do we have a chance to change the TCP settings in Shiny? Or the issue lies in somewhere else?

**Comment by erdalkucukk** [comment]
csrf_token|s:64:"1d42655502ba601c031ac3b3280b95a92e4a9bbbb2f088c9caa4dd039f5dbe43";valid|b:1;user|s:14:"dashboard.test";groups|s:392 <---- works fine

csrf_token|s:64:"e9a40398528bf800beee47c7351b8d8ffac0d6e6554ab96a27b0cfb83d28a361";valid|b:1;user|s:15:"dashboard.test2";groups|s:4338  <--- doesn't work.

do we have a limitation somewhere for the size of the token in the server?

**Comment by fereshtehRS** [comment]
@erdalkucukk 
Since this is a Shiny Server Pro issue I suggest that you open a support ticket, and our support and engineering teams will follow up with the issue.

We are aware of a character limit issue in certain scenarios. For example for LDAP groups when the list is more than 3000 characters. 
In case of proxy authentication, when I try with a long list of groups I get an error on the Nginx side:

```
upstream sent too big header while reading response header from upstream
```

What is the length of groups list in your case?

**Comment by erdalkucukk** [comment]
The length is 4338 characters for groups. So above the limit. I was looking for that information actually. 

For the upstream error, I managed to fix it after adding buffering in the nginx configuration. The length issue came as second step.

**Comment by fereshtehRS** [comment]
You might want to try adding `group_list` to you configuration and limit the number of groups. Please refer to this section of Admin Guide:

http://docs.rstudio.com/shiny-server/#group-authentication

This feature was meant as a workaround for LDAP groups issue, but it might help your situation too.

**Comment by erdalkucukk** [comment]
Thanks for letting me know, I tried some of it but appereantly I cannot use it location by location basis (according to the docs) and I made sure of it since when I implement it under locations, shiny server won't restart.  According to guide, it should be used in global level. 

Unfortunately I cannot implement it in global level due to our AD system is hierarchical. And even not sure since we have our NGINX before the ldap.

**Comment by fereshtehRS** [comment]
@erdalkucukk That is correct; `group_list` can only be defined at global (root) level. Does it seem to work when you define it at root level? If it does, it is very flexible, and you might be able to find a way to accommodate for your AD system, and still lower the list of groups. Here are all  the options you can use:

```
group_list include "Administrators";
group_list include_glob "*Administrato*";
group_list exclude "Administrators";
group_list exclude_glob "Administrato*";
group_list exclude_regex "regex_pattern";  
group_list include_regex "regex_pattern"; 
```

And you can define it multiple times.

**Comment by erdalkucukk** [comment]
Hi @fereshtehRS, I used the group_list feature under the root, had a successful shiny-server restart and tried with logging in and unfortunately had the same erroraneus output.

I used group_list exclude "xxxx"; multiple times for different groups I don't want to receive now. I can check the php sessions, but it doesn't change. Where can I check in Shiny, the reduced group_list, to check if it works? 

And do you have other suggestions?

**Comment by fereshtehRS** [comment]
@erdalkucukk Since you have a support ticket open for this, let's continue the conversation there.

**Comment by klopianoRTA** [comment]
Can you post what you ultimately did to resolve this issue? We are running into a similar issue.

Thanks!
